File: 2379. Minimum Recolors to Get K Consecutive Black Blocks/minimum_recolors.rs
use std::cmp;


impl Solution {
    pub fn minimum_recolors(blocks: String, k: i32) -> i32 {
        let mut left = 0;
        let mut right = k as usize;
        
        let mut white_count = 0;
        let mut black_count = 0;
        let mut chrs = Vec::new();
        for ch in blocks.chars() {
            chrs.push(ch);
        }
        
        for i in 0..right {
            if chrs[i] == 'W' {
                white_count += 1;
            } else {
                black_count += 1;
            }
        }
        
        let mut min_ops = white_count;
        while right < blocks.len() {
            if chrs[left] == 'W' { white_count -= 1 } else { black_count -= 1 }
            left += 1;
            if chrs[right] == 'W' { white_count += 1 } else { black_count += 1 }
            right += 1;
            min_ops = cmp::min(min_ops, white_count); 
        }
        
        return min_ops;
    }
}

File: 2379. Minimum Recolors to Get K Consecutive Black Blocks/MinimumRecolors.py
'''
https://leetcode.com/problems/minimum-recolors-to-get-k-consecutive-black-blocks/
'''


from collections import Counter


class Solution:
    # O(n) time,
    # O(1) space,
    # Approach: sliding window, hashmap, two pointers
    def minimumRecolors(self, blocks: str, k: int) -> int:
        n = len(blocks)
        
        l, r = 0, k
        window = Counter(blocks[l:r])
        # print(window)
        min_ops = window['W']
        
        while r < n:
            window[blocks[l]] -=1
            window[blocks[r]] +=1
            
            whites = window['W']
            if whites == 0:
                return 0
            
            min_ops = min(min_ops, whites)
            l +=1
            r +=1
            
        return min_ops